modules/gapi/src/compiler/passes/pattern_matching.cpp

```diff
@@ -0,0 +1,371 @@
+#include <unordered_set>
+
+#include "pattern_matching.hpp"
+
+cv::gapi::SubgraphMatch cv::gapi::findMatches(cv::gimpl::GModel::Graph patternGraph, cv::gimpl::GModel::Graph compGraph) {
+    using Graph = cv::gimpl::GModel::Graph;
+    using Metadata = typename Graph::MetadataT;
+    using VisitedMatchings = std::list<std::pair<ade::NodeHandle, ade::NodeHandle>>;
+
+    //TODO: Possibly, we may add N^2 check whether this graph may match or not at all.
+    //      Check that all pattern OP nodes exist in computational graph.
+
+    std::unordered_set<ade::NodeHandle, ade::HandleHasher<ade::Node>> firstPatternOpNodes;
+    std::unordered_set<ade::NodeHandle, ade::HandleHasher<ade::Node>> lastPatternOpNodes;
+
+    auto firstPatternDataNodes = patternGraph.metadata().get<cv::gimpl::Protocol>().in_nhs;
+    auto lastPatternDataNodes = patternGraph.metadata().get<cv::gimpl::Protocol>().out_nhs;
+
+    for (auto node : firstPatternDataNodes) {
```

Comment:
Why not `auto &`? (same for other cycles below)

Reply:
Done